src/jobserver.h

```diff
@@ -0,0 +1,271 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <assert.h>
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+#include <utility>
+
+/// Jobserver provides types related to managing a pool of "job slots"
+/// using the GNU Make jobserver ptocol described at:
+///
+/// https://www.gnu.org/software/make/manual/html_node/Job-Slots.html
+///
+struct Jobserver {
+  /// A Jobserver::Slot models a single job slot that can be acquired from.
+  /// or released to a jobserver pool. This class is move-only, and can
+  /// wrap three types of values:
+  ///
+  /// - An "invalid" value (the default), used to indicate errors, e.g.
+  ///   that no slot could be acquired from the pool.
+  ///
+  /// - The "implicit" value, used to model the job slot that is implicitly
+  ///   assigned to a jobserver client by the parent process that spawned
+  ///   it.
+  ///
+  /// - The "explicit" values, which correspond to an actual byte read from
+  ///   the slot pool's pipe (for Posix), or a semaphore decrement operation
+  ///   (for Windows).
+  ///
+  /// Use IsValid(), IsImplicit(), HasValue() to test for categories.
+  ///
+  /// TECHNICAL NOTE: This design complies with the requirements laid out
+  /// on https://www.gnu.org/software/make/manual/html_node/POSIX-Jobserver.html
+  /// which requires clients to write back the exact token values they
+  /// received from a Posix pipe.
+  ///
+  /// Note that *currently* all pool implementations write the same token
+  /// values to the pipe ('+' for GNU Make, and '|' for the Rust jobserver),
+  /// and do not care about the values written back by clients.
+  ///
+  struct Slot {
+    /// Default constructor creates invalid instance.
+    Slot() = default;
+
+    /// Move operations are allowed.
+    Slot(Slot&& o) noexcept : value_(o.value_) { o.value_ = -1; }
+
+    Slot& operator=(Slot&& o) noexcept {
+      if (this != &o) {
+        this->value_ = o.value_;
+        o.value_ = -1;
+      }
+      return *this;
+    }
+
+    /// Copy operations are disallowed.
+    Slot(const Slot&) = delete;
+    Slot& operator=(const Slot&) = delete;
+
+    /// Return true iff this instance is valid, i.e. that it is either
+    /// implicit or explicit job slot.
+    bool IsValid() const { return value_ >= 0; }
+
+    /// Return true iff this instance represents an implicit job slot.
+    bool IsImplicit() const { return value_ == kImplicitValue; }
+
+    /// Return true iff this instance represents an explicit job slot
+    bool IsExplicit() const { return IsValid() && !IsImplicit(); }
+
+    /// Return value of an explicit slot. It is a runtime error to call
+    /// this from an invalid instance.
+    uint8_t GetExplicitValue() const {
+      assert(IsExplicit());
+      return static_cast<uint8_t>(value_);
+    }
```

Comment:
Move implementation to jobserver.cc to get rid of the dependency on `assert.h` in a header file.